### PR TITLE
feat: ingress metrics by login

### LINKED ISSUE #20612: Add configurable, opt-in 'points written per login/measurement' ingress metric
__Proposal:__
As an operator, I want to see the number of points written per database login or per measurement, so I can understand the usage of a large influxdb instance.

__Current behavior:__
Currently the total number of points written for the whole instance is calculated - see https://github.com/influxdata/influxdb/blob/5b29ef7060ce37f6cc296025032f9cec822f009b/v1/coordinator/points_writer.go#L287

__Desired behavior:__
We want to add an opt-in feature to break down write volume metrics more granularly. By default we would break by database/retention policy, and optionally also by login user and/or measurement name.

__Alternatives considered:__
Doing a continuous query with `select count(*) from /.*/` or similar would give some of this granularity but would not provide the ability to break down by login user.

__Use case:__
Operators of large instances can sometimes have difficulty figuring out which database users are contributing to load. This feature would give them more visibility into where write load is coming from.

### PR DESCRIPTION
**Pull request by lesam**
After turning on authentication and both forms of ingress metrics:

"ingress": {"name":"ingress","tags":{"db":"_internal","login":"_systemuser_monitor","measurement":"cq","rp":"monitor"},"values":{"pointsWritten":38,"valuesWritten":76}},
"ingress:1": {"name":"ingress","tags":{"db":"_internal","login":"_systemuser_monitor","measurement":"database","rp":"monitor"},"values":{"pointsWritten":76,"valuesWritten":152}},
"ingress:2": {"name":"ingress","tags":{"db":"_internal","login":"_systemuser_monitor","measurement":"httpd","rp":"monitor"},"values":{"pointsWritten":38,"valuesWritten":874}},
"ingress:3": {"name":"ingress","tags":{"db":"_internal","login":"_systemuser_monitor","measurement":"ingress","rp":"monitor"},"values":{"pointsWritten":534,"valuesWritten":1068}},
"ingress:4": {"name":"ingress","tags":{"db":"_internal","login":"_systemuser_monitor","measurement":"localStore","rp":"monitor"},"values":{"pointsWritten":38,"valuesWritten":76}},
"ingress:5": {"name":"ingress","tags":{"db":"_internal","login":"_systemuser_monitor","measurement":"queryExecutor","rp":"monitor"},"values":{"pointsWritten":38,"valuesWritten":190}},
"ingress:6": {"name":"ingress","tags":{"db":"_internal","login":"_systemuser_monitor","measurement":"runtime","rp":"monitor"},"values":{"pointsWritten":38,"valuesWritten":570}},
"ingress:7": {"name":"ingress","tags":{"db":"_internal","login":"_systemuser_monitor","measurement":"shard","rp":"monitor"},"values":{"pointsWritten":76,"valuesWritten":836}},
"ingress:8": {"name":"ingress","tags":{"db":"_internal","login":"_systemuser_monitor","measurement":"subscriber","rp":"monitor"},"values":{"pointsWritten":38,"valuesWritten":114}},
"ingress:9": {"name":"ingress","tags":{"db":"_internal","login":"_systemuser_monitor","measurement":"tsm1_cache","rp":"monitor"},"values":{"pointsWritten":76,"valuesWritten":684}},
"ingress:10": {"name":"ingress","tags":{"db":"_internal","login":"_systemuser_monitor","measurement":"tsm1_engine","rp":"monitor"},"values":{"pointsWritten":76,"valuesWritten":2204}},
"ingress:11": {"name":"ingress","tags":{"db":"_internal","login":"_systemuser_monitor","measurement":"tsm1_filestore","rp":"monitor"},"values":{"pointsWritten":76,"valuesWritten":152}},
"ingress:12": {"name":"ingress","tags":{"db":"_internal","login":"_systemuser_monitor","measurement":"tsm1_wal","rp":"monitor"},"values":{"pointsWritten":76,"valuesWritten":304}},
"ingress:13": {"name":"ingress","tags":{"db":"_internal","login":"_systemuser_monitor","measurement":"write","rp":"monitor"},"values":{"pointsWritten":38,"valuesWritten":342}},
"ingress:14": {"name":"ingress","tags":{"db":"telegraf","login":"admin","measurement":"cpu","rp":"autogen"},"values":{"pointsWritten":1,"valuesWritten":1}},
"ingress:15": {"name":"ingress","tags":{"db":"telegraf","login":"telegraf","measurement":"cpu","rp":"autogen"},"values":{"pointsWritten":1316,"valuesWritten":13160}},
"ingress:16": {"name":"ingress","tags":{"db":"telegraf","login":"telegraf","measurement":"disk","rp":"autogen"},"values":{"pointsWritten":642,"valuesWritten":4494}},
"ingress:17": {"name":"ingress","tags":{"db":"telegraf","login":"telegraf","measurement":"diskio","rp":"autogen"},"values":{"pointsWritten":214,"valuesWritten":2354}},
"ingress:18": {"name":"ingress","tags":{"db":"telegraf","login":"telegraf","measurement":"mem","rp":"autogen"},"values":{"pointsWritten":107,"valuesWritten":963}},
"ingress:19": {"name":"ingress","tags":{"db":"telegraf","login":"telegraf","measurement":"processes","rp":"autogen"},"values":{"pointsWritten":107,"valuesWritten":856}},
"ingress:20": {"name":"ingress","tags":{"db":"telegraf","login":"telegraf","measurement":"swap","rp":"autogen"},"values":{"pointsWritten":214,"valuesWritten":642}},
"ingress:21": {"name":"ingress","tags":{"db":"telegraf","login":"telegraf","measurement":"system","rp":"autogen"},"values":{"pointsWritten":321,"valuesWritten":749}},

Only by login:

"ingress": {"name":"ingress","tags":{"login":"_systemuser_monitor"},"values":{"pointsWritten":42,"valuesWritten":354}},
"ingress:1": {"name":"ingress","tags":{"login":"admin"},"values":{"pointsWritten":1,"valuesWritten":1}},
"ingress:2": {"name":"ingress","tags":{"login":"telegraf"},"values":{"pointsWritten":3547,"valuesWritten":28246}},

Notice writes by users 'telegraf', '_systemuser_monitor', and 'admin'.

Closes #20612

- [x] [Well-formatted commit messages](https://www.conventionalcommits.org/en/v1.0.0-beta.3/)
- [x] Rebased/mergeable
- [x] Tests pass
- [ ] Documentation updated or issue created (provide link to issue/pr)

### CODE REVIEW

File: coordinator/points_writer.go [hunk 1]
```diff
@@ -280,15 +280,26 @@ func (l sgList) Append(sgi meta.ShardGroupInfo) sgList {
 // a cluster structure for information. This is to avoid a circular dependency.
 // It is used for 'SELECT INTO' statements
 func (w *PointsWriter) WritePointsInto(p *IntoWriteRequest) error {
-	return w.WritePointsPrivileged(p.Database, p.RetentionPolicy, models.ConsistencyLevelOne, p.Points)
+	// TODO: assign the correct original user for select into statements
```
Comment (lesam) [inline coordinator/points_writer.go:283]: Note the gap here

File: tsdb/config.go [hunk 2]
```diff
@@ -109,6 +109,7 @@ type Config struct {
 
 	// Options for ingress metrics
 	IngressMetricByMeasurement bool `toml:"ingress-metric-by-measurement-enabled"`
+	IngressMetricByLogin       bool `toml:"ingress-metric-by-login-enabled"`
```
Comment (davidby-influx) [inline tsdb/config.go:112]: Do we want to put some description and examples of this in `etc/config.sample.toml`?

File: cmd/influxd/run/server.go [hunk 3]
```diff
@@ -645,7 +645,10 @@ func (s *Server) stopProfile() error {
 type monitorPointsWriter coordinator.PointsWriter
 
 func (pw *monitorPointsWriter) WritePoints(database, retentionPolicy string, points models.Points) error {
-	return (*coordinator.PointsWriter)(pw).WritePointsPrivileged(database, retentionPolicy, models.ConsistencyLevelAny, points)
+	writeCtx := tsdb.WriteContext{
```
Comment (davidby-influx) [inline cmd/influxd/run/server.go:648]: Is it okay to have something named` WriteContext` which doesn't have a context.Context inside of it?

**Review by davidby-influx — APPROVED** [review]
A few questions, nothing blocking merge.

File: tsdb/config.go (hunk 2 above)
Comment (lesam) [reply tsdb/config.go:112]: Yes that would be a good idea - will fix

File: cmd/influxd/run/server.go (hunk 3 above)
Comment (lesam) [reply cmd/influxd/run/server.go:648]: `context.Context` does two things:

* cancellation
* request-scoped values

I really dislike using the request-scoped-values and I think they are mostly an antipattern. Here we need request scoped values but not cancellation. See e.g. https://lestrrat.medium.com/alternative-to-using-context-value-f2efe6bd2788 - I'm not alone.

As written this is a bit ugly though, I should probably either move `writeCtx` to be the first argument, or rename it to something like `WriteDetails`.

Comment (lesam) [reply cmd/influxd/run/server.go:648]: Note we could put a `context.Context` inside `tsdb.WriteContext` as a member, if we want cancellation later.

Comment (dgnorton) [reply cmd/influxd/run/server.go:648]: I'm also not a fan of passing request-scoped-values through `context.Context` even though that's partially what it was designed for and is common practice in Go. I find that code harder to follow.

Naming is tough. Anything with `Context` in the name in Go will initially make most people think of `context.Context` but I don't think that should discourage using `Context` in a type name where it best describes the type. `WriteContext` seems okay to me here.

**Review by davidby-influx — APPROVED** [review]
Looks good to me!

File: etc/config.sample.toml [hunk 4]
```diff
@@ -67,6 +67,12 @@
   # log any sensitive data contained within a query.
   # query-log-enabled = true
 
+  # It is possible to collect statistics of points written per-measurement and/or per-login.
```
Comment (lesam) [inline etc/config.sample.toml:70]: Config change

**Pull request merged**